### ISSUE TITLE
Enabling Referential Integrity in OpenDJ By Default

### ISSUE THREAD

**Issue by afroDC** [GluuFederation/community-edition-setup#472]
Currently Referential integrity is disabled in our edition of OpenDJ. We should enable this by default.

>The Referential Integrity Plugin maintains referential integrity for DN valued attributes.

> The values of these attributes can reference entries that have been deleted by a delete operation or renamed by a modify DN operation. The referential integrity plug-in either removes stale references to deleted entries or updates references to renamed entries. The plug-in allows the scope of this referential check to be limited to a set of base DNs if desired. The plug-in also can be configured to perform the referential checking in the background mode specified intervals.

```
/ # /opt/opendj/bin/ldapsearch -p 1636 -Z -X -D 'cn=directory manager' -w secret
 -b "cn=Referential Integrity,cn=Plugins,cn=config" "objectclass=*"
dn: cn=Referential Integrity,cn=Plugins,cn=config
objectClass: ds-cfg-plugin
objectClass: ds-cfg-referential-integrity-plugin
objectClass: top
ds-cfg-attribute-type: member
ds-cfg-attribute-type: uniqueMember
ds-cfg-enabled: false
cn: Referential Integrity
ds-cfg-java-class: org.opends.server.plugins.ReferentialIntegrityPlugin
ds-cfg-invoke-for-internal-operations: true
ds-cfg-plugin-type: postOperationDelete
ds-cfg-plugin-type: postOperationModifyDN
ds-cfg-plugin-type: subordinateModifyDN
ds-cfg-plugin-type: subordinateDelete
```

This would be a great feature for Gluu Casa when dealing with `oxAssociatedClients`.

**Comment by jgomer2001** [comment]
... and  for instance `associatedPerson` (`oxAuthClient`) or group memberships

**Comment by yurem** [comment]
It's good for RDBMS when we can enable cascade removal. But I don't think this is good for LDAP. And in Couchbase this will not work too. I offer to allow admin to enable this if needed.

@yuriyz @jgomer2001 @mbaser What do you think

**Comment by yuriyz** [comment]
I guess we can think on supporting referential integrity on persistence layer maybe? During entry load we can check whether references are stale (1 level only). The advantage 
is that it helps to keep data consistent. Also it's indeed not for 4.0. Lets schedule it on 4.1.

re: opendj plugin I agree, admin can enable it if needed. 

re: couchbase
http://codehardblog.azurewebsites.net/validating-couchbase-referential-integrity-with-n1ql/

**Comment by yurem** [comment]
I'm moving it to 4.1. But I think we need to close it. Thanks to LDAP we wrote code wich not depends on transactions, etc. As result each operation in our case is atomic.

Example:
1. Load client
2. Then load scopes specified in client, etc...
3. If one or more scopes are missing we will get list of actual scopes
4. Do specific action with hcope

For me we invited another way to do refferal integrity. Our code do it at runtime.
In DB server on it on write/modify/delete.
